api/ws/rpc/wsRPC.js

```diff
@@ -0,0 +1,188 @@
+ * Example:
+ * methodA registered on WS server can be called by a client by simply:
+ * sampleClientStub.methodA(exampleArg, cb);
+ *
+ * @typedef {Object} clientStub
+ * @property {function} procedure - procedure that will be called with argument and callback
+ */
+
+
+/**
+ * @param {ConnectionState} connectionState
+ * @returns {clientStub}
+ */
+var ClientRPCStub = function (connectionState) {
+	try {
+		var wsServer = wsRPC.getServer();
+	} catch (wsServerNotInitializedException) {
+		return {};
+	}
+
+	return _.reduce(Object.assign({}, wsServer.endpoints.rpc, wsServer.endpoints.event),
+		function (availableCalls, procedureHandler, procedureName) {
+			availableCalls[procedureName] = this.sendAfterSocketReadyCb(connectionState)(procedureName);
+			return availableCalls;
+		}.bind(this), {});
+};
+
+
+/**
+ * @param {ConnectionState} connectionState
+ */
+ClientRPCStub.prototype.initializeNewConnection = function (connectionState) {
+
+	var options = {
+		hostname: connectionState.ip,
+		port: connectionState.port,
+		protocol: 'http',
+		autoReconnect: true,
+		query: constants.getConst('headers')
+	};
+
+	var clientSocket = wsRPC.scClient.connect(options);
+	wsRPC.wampClient.upgradeToWAMP(clientSocket);
+
+
+	clientSocket.on('connect', function () {
+		return connectionState.resolve(clientSocket);
+	});
+
+	clientSocket.on('error', function () {
+		clientSocket.disconnect();
+	});
+
+	clientSocket.on('connectAbort', function (err, data) {
+		connectionState.reject('Connection rejected by failed handshake procedure');
+	});
+
+	clientSocket.on('disconnect', function () {
+		connectionState.reject('Connection disconnected');
+	});
+};
+
+/**
+ * @param {ConnectionState} connectionState
+ * @returns {function} function to be called with procedure, to be then called with optional argument and/or callback
+ */
+ClientRPCStub.prototype.sendAfterSocketReadyCb = function (connectionState) {
+	return function (procedureName) {
+		/**
+		 * @param {object} data [data={}] argument passed to procedure
+		 * @param {function} cb [cb=function(){}] cb
+		 */
+		return function (data, cb) {
+			cb = _.isFunction(cb) ? cb : _.isFunction(data) ? data : function () {};
+			data = (data && !_.isFunction(data)) ? data : {};
+
+			if (connectionState.status === ConnectionState.STATUS.NEW || connectionState.status === ConnectionState.STATUS.DISCONNECTED) {
+				connectionState.reconnect();
+				ClientRPCStub.prototype.initializeNewConnection(connectionState);
+			}
+
+			connectionState.socketDefer.promise.then(function (socket) {
+				return socket.wampSend(procedureName, data)
+					.then(function (res) {
+						return setImmediate(cb, null, res);
+					})
+					.catch(function (err) {
+						return setImmediate(cb, err);
+					});
+			}).catch(function (err) {
+				return setImmediate(cb, err);
+			});
+		};
+	};
+};
+
+module.exports = {
+	wsRPC: wsRPC,
+	ConnectionState: ConnectionState,
+	ClientRPCStub: ClientRPCStub
```

Comment:
Actually, this so I do not know if there is a reason to change it into a class with initialization.
The reason why wsRPC has a different structure than ClientRPCStub is that there can be only one WS server instance across the app, while many clients